# **Updated population density of the LA city neighborhood councils.**
    
### Here is a notebook to evaluate the populataion, area(in square miles) and population density of the 99 Neighborhood councils (NCs). 

In [1]:
# Importing the necessary packages.
import os
import pandas as pd
import geopandas as gpd
import matplotlib.pyplot as plt 
import folium
from shapely.geometry import Point, Polygon 
from pyproj import Geod
import numpy as np
import webbrowser

In [2]:
# Path to the directory.
dir = 'C:/Users/AdithiPriya/Desktop/Hack for LA/Geospatial analysis'

# Accessing the Census 2020 data.
census_2020 = gpd.read_file(os.path.join(dir, 'tl_2020_06037_tract20/tl_2020_06037_tract20.shp'))

# Reading the LA NC shape file.
la_nc = gpd.read_file(os.path.join(dir, 'Neighborhood_Councils_(Certified)/Neighborhood_Councils_(Certified).shp'))

# ACS demographics data.
acs_demo = pd.read_csv(os.path.join(dir,'ACS_census_tract_LA.csv' ))
acs_demo['GEO_ID']= acs_demo['GEO_ID'].str.replace('1400000US','')

# Rename GEO_ID as GEOID20.

acs_demo= acs_demo.rename(columns={'GEO_ID':'GEOID20', 'Total population': 'population'})

### Defining a [function](https://github.com/hackforla/access-the-data-workshop-311-analysis/blob/main/notebooks/NC-population-density.ipynb) to compute the area in square miles (area (sq_miles)).

In [3]:
geod = Geod(ellps= 'WGS84') 

def area_sq_miles(geo):
    area_sq_meters= abs(geod.geometry_area_perimeter(geo)[0])
    return (area_sq_meters * 3.86102e-7)

In [4]:
# Taking a subset of the acs_demo dataframe.

acs_demo_subset=acs_demo[['GEOID20','population' ]]

# Making sure that the crs- coordinate reference system for census_2020 is the same as that of the la_nc using the to_crs()
# method. This will allow the spatial merging of both the geopandas dataframes.

census_2020 = census_2020.to_crs(la_nc.crs)

# Spatial overlap of the la_nc and census_2020 data.

census_NC = gpd.overlay(census_2020, la_nc, how='intersection')

census_NC= census_NC[['TRACTCE20','GEOID20', 'NAME', 'NC_ID', 'INTPTLAT20', 'INTPTLON20', 'geometry']]
print('Number of rows in census_NC: ', census_NC.shape[0])  

Number of rows in census_NC:  2108


In [5]:
# Let us take a look at the spatially merged data:
census_NC.head()

TRACTCE20      GEOID20                NAME  NC_ID   INTPTLAT20  \
0    101110  06037101110  SUNLAND-TUJUNGA NC     10  +34.2594737   
1    101122  06037101122  SUNLAND-TUJUNGA NC     10  +34.2677213   
2    101220  06037101220  SUNLAND-TUJUNGA NC     10  +34.2516083   
3    101221  06037101221  SUNLAND-TUJUNGA NC     10  +34.2543290   
4    101222  06037101222  SUNLAND-TUJUNGA NC     10  +34.2513519   

     INTPTLON20                                           geometry  
0  -118.2929869  POLYGON ((-118.30229 34.25870, -118.30091 34.2...  
1  -118.2901465  POLYGON ((-118.30334 34.27371, -118.30330 34.2...  
2  -118.2816328  POLYGON ((-118.28592 34.25227, -118.28592 34.2...  
3  -118.2925767  POLYGON ((-118.29945 34.25598, -118.29792 34.2...  
4  -118.2885261  POLYGON ((-118.29434 34.25233, -118.29318 34.2...

In [6]:
# Merging the census_NC and acs_demo_subset.
census = pd.merge(census_NC, acs_demo_subset, on='GEOID20')
census.shape[0]

2108

### Displaying the duplicate entries- census tracts intersecting more than 1 Neighborhood councils

In [7]:
df_duplicate = census_NC[census_NC.GEOID20.duplicated(keep=False)]
df_duplicate.sort_values(by=['GEOID20']).head(15)

TRACTCE20      GEOID20                         NAME  NC_ID   INTPTLAT20  \
1     101122  06037101122           SUNLAND-TUJUNGA NC     10  +34.2677213   
17    101122  06037101122  FOOTHILL TRAILS DISTRICT NC      9  +34.2677213   
5     101300  06037101300           SUNLAND-TUJUNGA NC     10  +34.2487777   
18    101300  06037101300  FOOTHILL TRAILS DISTRICT NC      9  +34.2487777   
6     101400  06037101400           SUNLAND-TUJUNGA NC     10  +34.2428521   
19    101400  06037101400  FOOTHILL TRAILS DISTRICT NC      9  +34.2428521   
42    102103  06037102103           SUN VALLEY AREA NC      8  +34.2250792   
20    102103  06037102103  FOOTHILL TRAILS DISTRICT NC      9  +34.2250792   
43    102104  06037102104           SUN VALLEY AREA NC      8  +34.2161873   
21    102104  06037102104  FOOTHILL TRAILS DISTRICT NC      9  +34.2161873   
45    102107  06037102107           SUN VALLEY AREA NC      8  +34.2409052   
22    102107  06037102107  FOOTHILL TRAILS DISTRICT NC      9  +34.2409052   
23    103101  06037103101  FOOTHILL TRAILS DISTRICT NC      9  +34.2732407   
7     103101  06037103101           SUNLAND-TUJUNGA NC     10  +34.2732407   
24    103201  06037103201  FOOTHILL TRAILS DISTRICT NC      9  +34.2656446   

      INTPTLON20                                           geometry  
1   -118.2901465  POLYGON ((-118.30334 34.27371, -118.30330 34.2...  
17  -118.2901465  POLYGON ((-118.29785 34.27778, -118.29783 34.2...  
5   -118.2709990  POLYGON ((-118.27822 34.25068, -118.27822 34.2...  
18  -118.2709990  POLYGON ((-118.26682 34.23124, -118.26695 34.2...  
6   -118.2941612  POLYGON ((-118.32227 34.24961, -118.32212 34.2...  
19  -118.2941612  POLYGON ((-118.32238 34.24963, -118.32227 34.2...  
42  -118.3541880  POLYGON ((-118.36533 34.22870, -118.36396 34.2...  
20  -118.3541880  POLYGON ((-118.35739 34.22856, -118.35546 34.2...  
43  -118.3453981  POLYGON ((-118.35620 34.21971, -118.35594 34.2...  
21  -118.3453981  MULTIPOLYGON (((-118.34413 34.21387, -118.3441...  
45  -118.3395988  POLYGON ((-118.35739 34.22856, -118.35848 34.2...  
22  -118.3395988  POLYGON ((-118.36789 34.23939, -118.36788 34.2...  
23  -118.3078874  POLYGON ((-118.31534 34.27876, -118.31533 34.2...  
7   -118.3078874  POLYGON ((-118.31760 34.27387, -118.31758 34.2...  
24  -118.3236267  MULTIPOLYGON (((-118.33294 34.26501, -118.3329...

In [8]:
df_duplicate.shape[0]

1475

### Out of 2108 rows, 1475 rows are duplicate. 

### Here is a map of NCs and census tracts. Zoom in and toggle between NC and census tracts to see that some of the NCs share the census tracts with their neighbors. 

### Here is a link to the [map](https://exquisite-squirrel-d3afcc.netlify.app/). I am using netlify to deploy the interactive map. 

In [9]:
m= la_nc.explore(
    column= 'NAME', # make choropleth based on 'NC name' column
    name='NC Regions', 
    tooltip='NAME', # show 'NC name' value in tooltip (on hover)
    color="black", # use red color on all points
    popup=True, # show all values in popup (on click)
    tiles="openstreetmap", # use "openstreetmap" tiles
    cmap="Set1", # use "Set1" matplotlib colormap
    style_kwds=dict(color="black"), # use black outline
    legend=False
     )
census_NC.explore(
    m=m,
    column= 'TRACTCE20', # make choropleth based on 'NC name' column
    name='Census tracts', 
    tooltip='TRACTCE20', # show 'NC name' value in tooltip (on hover)
    color="red", # use red color on all points
    popup=True, # show all values in popup (on click)
    tiles="openstreetmap", # use "openstreetmap" tiles
    cmap="Set1", # use "Set1" matplotlib colormap
    style_kwds=dict(color="black"), # use black outline
    legend=False
     )
folium.TileLayer('Stamen Terrain').add_to(m)
folium.TileLayer('Stamen Toner').add_to(m)
folium.TileLayer('Stamen Water Color').add_to(m)
folium.TileLayer('cartodbpositron').add_to(m)
folium.TileLayer('cartodbdark_matter').add_to(m)
folium.LayerControl().add_to(m)
m.save('census_plus_NC.html')
webbrowser.open('census_plus_NC.html')

True

### Before moving forward, for sanity check, let us try to get the population of distinct census tracts without worrying about the proportion of NC areas intersected by census tracts. To do this, I will retain the first row of the  duplicate data for the entire dataframe.

In [10]:
test =  census.groupby('TRACTCE20').first()
test.population.sum()

4534712

<div style="text-align: justify">  
    
### This number exactly matches with the one that I get using this [notebook](https://github.com/priyakalyan/Updated_NC_pop/blob/main/NC_pop_recent_no_filter.ipynb) where the population of NCs have been calculated at the tract level. This implies that the census tracts that is being used in the calculation is wrong. This issue will be addressed here. But before going ahead, the next section will outline the methodology to calculate the area of census tracts intersecting multiple NCs.

### So here is a method to take care of cases where the census tract intersects more than 1 NCs. 

- **Find the area for each entry in census dataframe.** 
- **Group the census dataframe by 'TRACTCE20' and then find sum of the area- this gives you the total area of each census       tract.** 
- **Next, find the percentage of area of the census tract intersecting different NCs.** 
- **Use this information to find the percentage of population for each census tract.**
- **Then find the total population by grouping them by NCs.** 

In [11]:
# Add the area column.
census = pd.merge(census_NC, acs_demo_subset, on='GEOID20')
census['area(sq_miles)']= census.apply(lambda x: area_sq_miles(x.geometry), axis=1)
census[['NC_ID','NAME', 'TRACTCE20', 'population','area(sq_miles)']].head(10)

NC_ID                         NAME TRACTCE20  population  area(sq_miles)
0     10           SUNLAND-TUJUNGA NC    101110        3923        0.441083
1     10           SUNLAND-TUJUNGA NC    101122        4119        1.020872
2      9  FOOTHILL TRAILS DISTRICT NC    101122        4119        0.000038
3     10           SUNLAND-TUJUNGA NC    101220        3775        0.269841
4     10           SUNLAND-TUJUNGA NC    101221        3787        0.136748
5     10           SUNLAND-TUJUNGA NC    101222        2717        0.114484
6     10           SUNLAND-TUJUNGA NC    101300        3741        0.993003
7      9  FOOTHILL TRAILS DISTRICT NC    101300        3741        0.002387
8     10           SUNLAND-TUJUNGA NC    101400        3246        2.414663
9      9  FOOTHILL TRAILS DISTRICT NC    101400        3246        0.021664

In [12]:
census_test=census.groupby('TRACTCE20', as_index= False).agg({'NC_ID':'first','NAME':'first','geometry':'first', 'INTPTLAT20':'first', 'INTPTLON20':'first', 'NAME':'first','GEOID20':'first','area(sq_miles)':sum})
census_test.rename(columns={'area(sq_miles)': 'total_area(sq_miles)'}, inplace=True)
census_test= census_test[['TRACTCE20','total_area(sq_miles)']]

In [13]:
# Let us add the total_area to census dataframe.
census_perc= pd.merge(census, census_test, on='TRACTCE20')
census_perc.head()

TRACTCE20      GEOID20                         NAME  NC_ID   INTPTLAT20  \
0    101110  06037101110           SUNLAND-TUJUNGA NC     10  +34.2594737   
1    101122  06037101122           SUNLAND-TUJUNGA NC     10  +34.2677213   
2    101122  06037101122  FOOTHILL TRAILS DISTRICT NC      9  +34.2677213   
3    101220  06037101220           SUNLAND-TUJUNGA NC     10  +34.2516083   
4    101221  06037101221           SUNLAND-TUJUNGA NC     10  +34.2543290   

     INTPTLON20                                           geometry  \
0  -118.2929869  POLYGON ((-118.30229 34.25870, -118.30091 34.2...   
1  -118.2901465  POLYGON ((-118.30334 34.27371, -118.30330 34.2...   
2  -118.2901465  POLYGON ((-118.29785 34.27778, -118.29783 34.2...   
3  -118.2816328  POLYGON ((-118.28592 34.25227, -118.28592 34.2...   
4  -118.2925767  POLYGON ((-118.29945 34.25598, -118.29792 34.2...   

   population  area(sq_miles)  total_area(sq_miles)  
0        3923        0.441083              0.441083  
1        4119        1.020872              1.020910  
2        4119        0.000038              1.020910  
3        3775        0.269841              0.269841  
4        3787        0.136748              0.136748

In [14]:
# Evaluating the percentage of the intersecting areas (area_perc):
census_perc['area_perc'] = census_perc['area(sq_miles)']/census_perc['total_area(sq_miles)']

# Adding the percentage of population column.
census_perc['total_population'] = census_perc['population']*census_perc['area_perc'] 

# Rounding the population number:
census_perc['total_population'] = census_perc['total_population'].apply(np.floor)

# Converting the population to integer type.

census_perc['total_population']  = census_perc['total_population'].astype(int)

census_perc[['TRACTCE20', 'NAME', 'NC_ID', 'population', 'area(sq_miles)', 'area_perc', 'total_area(sq_miles)', 'total_population']].head(15)

TRACTCE20                         NAME  NC_ID  population  area(sq_miles)  \
0     101110           SUNLAND-TUJUNGA NC     10        3923        0.441083   
1     101122           SUNLAND-TUJUNGA NC     10        4119        1.020872   
2     101122  FOOTHILL TRAILS DISTRICT NC      9        4119        0.000038   
3     101220           SUNLAND-TUJUNGA NC     10        3775        0.269841   
4     101221           SUNLAND-TUJUNGA NC     10        3787        0.136748   
5     101222           SUNLAND-TUJUNGA NC     10        2717        0.114484   
6     101300           SUNLAND-TUJUNGA NC     10        3741        0.993003   
7     101300  FOOTHILL TRAILS DISTRICT NC      9        3741        0.002387   
8     101400           SUNLAND-TUJUNGA NC     10        3246        2.414663   
9     101400  FOOTHILL TRAILS DISTRICT NC      9        3246        0.021664   
10    103101           SUNLAND-TUJUNGA NC     10        2484        0.607303   
11    103101  FOOTHILL TRAILS DISTRICT NC      9        2484        0.176436   
12    103102           SUNLAND-TUJUNGA NC     10        4716        0.479162   
13    103201           SUNLAND-TUJUNGA NC     10        3961        0.548078   
14    103201  FOOTHILL TRAILS DISTRICT NC      9        3961        0.206678   

    area_perc  total_area(sq_miles)  total_population  
0    1.000000              0.441083              3923  
1    0.999963              1.020910              4118  
2    0.000037              1.020910                 0  
3    1.000000              0.269841              3775  
4    1.000000              0.136748              3787  
5    1.000000              0.114484              2717  
6    0.997602              0.995390              3732  
7    0.002398              0.995390                 8  
8    0.991108              2.436327              3217  
9    0.008892              2.436327                28  
10   0.774879              0.783739              1924  
11   0.225121              0.783739               559  
12   1.000000              0.479162              4716  
13   0.726166              0.754756              2876  
14   0.273834              0.754756              1084

<div style="text-align: justify">  
    
### I will be incorporating the area filter and the population filter to skim out the values that were causing the population of the NCs to inflate. 

### Let us take a look at the NC- Atwater Village. 

In [15]:
census_perc[census_perc['NAME'] == 'ATWATER VILLAGE NC'][['TRACTCE20', 'NAME', 'NC_ID', 'population', 'area(sq_miles)', 'area_perc', 'total_area(sq_miles)', 'total_population']]

TRACTCE20                NAME  NC_ID  population  area(sq_miles)  \
811     189703  ATWATER VILLAGE NC     37        1880        0.043906   
833     311601  ATWATER VILLAGE NC     37        2608        0.000209   
837     311700  ATWATER VILLAGE NC     37        5966        0.004878   
841     980009  ATWATER VILLAGE NC     37           5        6.816197   
973     187102  ATWATER VILLAGE NC     37        4376        0.340444   
984     187101  ATWATER VILLAGE NC     37        3213        0.302714   
986     187200  ATWATER VILLAGE NC     37        3324        0.011649   
990     187300  ATWATER VILLAGE NC     37        3715        0.004740   
994     188100  ATWATER VILLAGE NC     37        3960        0.813095   
995     188201  ATWATER VILLAGE NC     37        3262        0.019628   
999     188202  ATWATER VILLAGE NC     37        2887        0.000121   
1002    188300  ATWATER VILLAGE NC     37        3822        0.356143   
1003    189202  ATWATER VILLAGE NC     37        2781        0.000332   
1006    189300  ATWATER VILLAGE NC     37        3212        0.022325   
1009    189400  ATWATER VILLAGE NC     37        3145        0.002692   
1012    301601  ATWATER VILLAGE NC     37        5657        0.002890   
1013    301602  ATWATER VILLAGE NC     37        4472        0.000794   
1016    301702  ATWATER VILLAGE NC     37        6075        0.005642   
1017    311801  ATWATER VILLAGE NC     37        2896        0.000050   

      area_perc  total_area(sq_miles)  total_population  
811    0.029314              1.497769                55  
833    0.149092              0.001404               388  
837    0.322874              0.015109              1926  
841    0.312739             21.795157                 1  
973    0.402493              0.845840              1761  
984    0.995004              0.304234              3196  
986    0.031713              0.367315               105  
990    0.010301              0.460138                38  
994    1.000000              0.813095              3960  
995    0.058066              0.338025               189  
999    0.000411              0.293574                 1  
1002   1.000000              0.356143              3822  
1003   0.000830              0.400121                 2  
1006   0.028684              0.778319                92  
1009   0.004774              0.564007                15  
1012   1.000000              0.002890              5657  
1013   0.331844              0.002393              1484  
1016   1.000000              0.005642              6075  
1017   0.332929              0.000149               964

### Look at the census tracts 301601, 301602, 301702, 311801. When I looked at the map, these were the tracts belonging to the neighboring city- Glendale. These tracts obviously do not belong to Atwater Village NC. The total area of these tracts are extremely small, but the area_perc for some of these tracts - 1 and so the whole population is taken into account and thus this will result in a higher population value than the expected one. So, I am adding the area filter- total_area(sq_miles) >= 0.01. 
    
### Similarly, some of the population value is so small, but the cumulative effect can result in an exaggerated population number. Therefore, I am going to add the population filter here. The [minimum population](https://www2.census.gov/geo/pdfs/education/CensusTracts.pdf) of a whole census tract is 1200. Since we are dealing with proportional areas and sometimes, a census tract intersects 2 NCs in 30% to 70% ratio and so I want to be reasonable with the population filter- 1000 (we do not want to undercount too). 

In [16]:
# Filtering out total_area < 0.01 square miles.
census_perc_filter = census_perc[census_perc['total_area(sq_miles)'] >= 0.01]
census_perc_filter.head()

TRACTCE20      GEOID20                         NAME  NC_ID   INTPTLAT20  \
0    101110  06037101110           SUNLAND-TUJUNGA NC     10  +34.2594737   
1    101122  06037101122           SUNLAND-TUJUNGA NC     10  +34.2677213   
2    101122  06037101122  FOOTHILL TRAILS DISTRICT NC      9  +34.2677213   
3    101220  06037101220           SUNLAND-TUJUNGA NC     10  +34.2516083   
4    101221  06037101221           SUNLAND-TUJUNGA NC     10  +34.2543290   

     INTPTLON20                                           geometry  \
0  -118.2929869  POLYGON ((-118.30229 34.25870, -118.30091 34.2...   
1  -118.2901465  POLYGON ((-118.30334 34.27371, -118.30330 34.2...   
2  -118.2901465  POLYGON ((-118.29785 34.27778, -118.29783 34.2...   
3  -118.2816328  POLYGON ((-118.28592 34.25227, -118.28592 34.2...   
4  -118.2925767  POLYGON ((-118.29945 34.25598, -118.29792 34.2...   

   population  area(sq_miles)  total_area(sq_miles)  area_perc  \
0        3923        0.441083              0.441083   1.000000   
1        4119        1.020872              1.020910   0.999963   
2        4119        0.000038              1.020910   0.000037   
3        3775        0.269841              0.269841   1.000000   
4        3787        0.136748              0.136748   1.000000   

   total_population  
0              3923  
1              4118  
2                 0  
3              3775  
4              3787

In [17]:
# Removing the rows with population value < 500.
census_filter = census_perc_filter[census_perc_filter['total_population'] >= 1000]
census_filter.shape[0]

1172

In [18]:
# Summming up the population of each NCs.
census_pop =census_filter.groupby('NAME', as_index= False).agg({'NC_ID':'first',  'NAME':'first', 'total_population' : sum})
census_pop['total_population'] = census_pop['total_population'].astype(int)
census_pop.head(50)

NC_ID                                               NAME  total_population
0       6                                          ARLETA NC             35585
1      42                                     ARROYO SECO NC             19479
2      46                      ARTS DISTRICT LITTLE TOKYO NC              4153
3      37                                 ATWATER VILLAGE NC             14665
4      64                           BEL AIR-BEVERLY CREST NC             24732
5      50                                   BOYLE HEIGHTS NC             87817
6      13                                     CANOGA PARK NC             54184
7     110                                 CENTRAL ALAMEDA NC             30219
8      32                               CENTRAL HOLLYWOOD NC             18204
9      95                               CENTRAL SAN PEDRO NC             29244
10     99                                      CHATSWORTH NC             41153
11     96                               COASTAL SAN PEDRO NC             28209
12     86  COMMUNITY AND NEIGHBORS FOR NINTH DISTRICT UNI...             48501
13     70                                         DEL REY NC             34235
14     52                               DOWNTOWN LOS ANGELES             50250
15     40                                      EAGLE ROCK NC             31518
16     34                                  EAST HOLLYWOOD NC             42741
17     44                                       ECHO PARK NC             42361
18     43                        ELYSIAN VALLEY RIVERSIDE NC              6926
19     81               EMPOWERMENT CONGRESS CENTRAL AREA NC             51301
20     77                 EMPOWERMENT CONGRESS NORTH AREA NC             78270
21     87            EMPOWERMENT CONGRESS SOUTHEAST AREA NDC             85528
22     84            EMPOWERMENT CONGRESS SOUTHWEST AREA NDC             30229
23     79                 EMPOWERMENT CONGRESS WEST AREA NDC             41645
24     18                                          ENCINO NC             50445
25      9                        FOOTHILL TRAILS DISTRICT NC             23882
26     39                                   GLASSELL PARK NC             22120
27      4                             GRANADA HILLS NORTH NC             26351
28    118                             GRANADA HILLS SOUTH NC             38583
29    102                            GREATER CYPRESS PARK NC             10620
30     28                             GREATER TOLUCA LAKE NC             12027
31     21                                GREATER VALLEY GLEN             44066
32    119                                GREATER WILSHIRE NC             45702
33     92                                     HARBOR CITY NC             25688
34     90                            HARBOR GATEWAY NORTH NC             37928
35     91                            HARBOR GATEWAY SOUTH NC             23167
36    126                                          HERMON NC              2649
37    128                         HISTORIC CULTURAL NORTH NC             26289
38     41                          HISTORIC HIGHLAND PARK NC             47327
39     29                            HOLLYWOOD HILLS WEST NC             36653
40     33                       HOLLYWOOD STUDIO DISTRICT NC             24286
41     30                                HOLLYWOOD UNITED NC             22884
42     48                                           LA-32 NC             45581
43     19                                     LAKE BALBOA NC             44375
44     47                                 LINCOLN HEIGHTS NC             31741
45     36                                       LOS FELIZ NC             34513
46     54                                  MACARTHUR PARK NC             27222
47     67                                       MAR VISTA NC             51439
48     73                                        MID CITY NC             26221
49     58                                   MID CITY W

In [19]:
census_pop.total_population.sum()

3825667

<div style="text-align: justify">  
    
### This number is close to the [Census Bureau](https://www.census.gov/quickfacts/losangelescitycalifornia?) value- 3,849,297. Now that we have the population of each NC, let us move on to getting the area and finally the updated population density of the neighborhood councils.      

In [20]:
# Grouping the original census dataframe by NAME and then summing up the total area- this gives the area of each neighborhood 
# council. Very important note here- groupby function works when trying to aggregate dataframes but for spatial data, we can
# aggregate the geometry features using dissolve function.

census_area = census.dissolve(by= 'NAME', as_index= False, aggfunc=({'NC_ID':'first','area(sq_miles)' : sum }))

# Gathering all the columns of interest. 
census_final = census_area.join(census_pop['total_population'])
census_final['pop_density']= census_final['total_population']/census_final['area(sq_miles)']

# Rearranging the columns.
census_final = census_final[['NAME', 'geometry', 'NC_ID', 'total_population', 'area(sq_miles)', 'pop_density']]
census_final.head(15)

NAME  \
0                                           ARLETA NC   
1                                      ARROYO SECO NC   
2                       ARTS DISTRICT LITTLE TOKYO NC   
3                                  ATWATER VILLAGE NC   
4                            BEL AIR-BEVERLY CREST NC   
5                                    BOYLE HEIGHTS NC   
6                                      CANOGA PARK NC   
7                                  CENTRAL ALAMEDA NC   
8                                CENTRAL HOLLYWOOD NC   
9                                CENTRAL SAN PEDRO NC   
10                                      CHATSWORTH NC   
11                               COASTAL SAN PEDRO NC   
12  COMMUNITY AND NEIGHBORS FOR NINTH DISTRICT UNI...   
13                                         DEL REY NC   
14                               DOWNTOWN LOS ANGELES   

                                             geometry  NC_ID  \
0   POLYGON ((-118.41010 34.23309, -118.41034 34.2...      6   
1   POLYGON ((-118.18576 34.09293, -118.18576 34.0...     42   
2   POLYGON ((-118.22877 34.04155, -118.22827 34.0...     46   
3   POLYGON ((-118.25399 34.10816, -118.25424 34.1...     37   
4   POLYGON ((-118.46573 34.07325, -118.46581 34.0...     64   
5   POLYGON ((-118.20504 34.01263, -118.20504 34.0...     50   
6   POLYGON ((-118.58846 34.19524, -118.58846 34.1...     13   
7   POLYGON ((-118.23777 33.98933, -118.23777 33.9...    110   
8   POLYGON ((-118.32445 34.08712, -118.32445 34.0...     32   
9   POLYGON ((-118.28794 33.73151, -118.28795 33.7...     95   
10  POLYGON ((-118.56231 34.23405, -118.56232 34.2...     99   
11  POLYGON ((-118.26232 33.70500, -118.26232 33.7...     96   
12  POLYGON ((-118.25640 33.96470, -118.25640 33.9...     86   
13  POLYGON ((-118.43312 33.97609, -118.43315 33.9...     70   
14  POLYGON ((-118.22419 34.01494, -118.22569 34.0...     52   

    total_population  area(sq_miles)   pop_density  
0              35585        3.284868  10833.007723  
1              19479        3.063327   6358.773341  
2               4153        0.879216   4723.528997  
3              14665        8.748450   1676.296980  
4              24732       17.038756   1451.514436  
5              87817        5.735881  15310.116685  
6              54184        3.689892  14684.440778  
7              30219        1.358014  22252.346382  
8              18204        1.229127  14810.517707  
9              29244        2.438025  11994.957379  
10             41153       11.394106   3611.779707  
11             28209        4.713196   5985.111214  
12             48501        2.550243  19018.185733  
13             34235        3.027096  11309.517225  
14             50250        5.030842   9988.388370

### Here is a plot of census_final dataframe with the population, population density and area popping up for each NCs. I am posting the link to this [map](https://ephemeral-salamander-ff5d33.netlify.app/).

In [21]:
m1 = census_final.explore(
    column= 'NAME', # make choropleth based on 'NC name' column
    name='NC Regions', 
    tooltip='NAME', # show 'NC name' value in tooltip (on hover)
    color="black", # use red color on all points
    popup=True, # show all values in popup (on click)
    tiles="openstreetmap", # use "openstreetmap" tiles
    cmap="Set1", # use "Set1" matplotlib colormap
    style_kwds=dict(color="black"), # use black outline
    legend=False
     )
folium.TileLayer('Stamen Terrain').add_to(m1)
folium.TileLayer('Stamen Toner').add_to(m1)
folium.TileLayer('Stamen Water Color').add_to(m1)
folium.TileLayer('cartodbpositron').add_to(m1)
folium.TileLayer('cartodbdark_matter').add_to(m1)
folium.LayerControl().add_to(m1)
m1.save('census_final.html')
webbrowser.open('census_final.html')

True

In [22]:
census_final.head()

NAME  \
0                      ARLETA NC   
1                 ARROYO SECO NC   
2  ARTS DISTRICT LITTLE TOKYO NC   
3             ATWATER VILLAGE NC   
4       BEL AIR-BEVERLY CREST NC   

                                            geometry  NC_ID  total_population  \
0  POLYGON ((-118.41010 34.23309, -118.41034 34.2...      6             35585   
1  POLYGON ((-118.18576 34.09293, -118.18576 34.0...     42             19479   
2  POLYGON ((-118.22877 34.04155, -118.22827 34.0...     46              4153   
3  POLYGON ((-118.25399 34.10816, -118.25424 34.1...     37             14665   
4  POLYGON ((-118.46573 34.07325, -118.46581 34.0...     64             24732   

   area(sq_miles)   pop_density  
0        3.284868  10833.007723  
1        3.063327   6358.773341  
2        0.879216   4723.528997  
3        8.748450   1676.296980  
4       17.038756   1451.514436

In [23]:
# Save the dataframe as a csv file after dropping the geometry column.
census_final= census_final.drop(['geometry'], axis=1)
census_final.to_csv('census_final.csv')